contracts/SafeProtocolMediator.sol

```diff
@@ -0,0 +1,168 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity ^0.8.18;
+import {ISafeProtocolMediator} from "./interfaces/Mediator.sol";
+import {ISafeProtocolModule} from "./interfaces/Components.sol";
+
+import {ISafe} from "./interfaces/Accounts.sol";
+import {SafeProtocolAction, SafeTransaction, SafeRootAccess} from "./DataTypes.sol";
+import {Ownable2Step} from "@openzeppelin/contracts/access/Ownable2Step.sol";
+
+/**
+ * @title TODO
+ * @notice TODO
+ */
+contract SafeProtocolMediator is ISafeProtocolMediator, Ownable2Step {
+    /**
+     * @notice Mapping of a mapping what stores information about modules that are enabled per Safe.
+     *         address (Safe address) => address (component address) => EnabledMoudleInfo
+     */
+    mapping(address => mapping(address => MoudleAccessInfo)) public enabledComponents;
+    struct MoudleAccessInfo {
+        bool enabled;
+        bool rootAddressGranted;
+        // TODO: Add deadline for validity
+    }
+
+    // Events
+    event ActionExecuted(address safe, bytes32 metaHash, uint256 index);
+    event RootAccessActionExecuted(address safe, bytes32 metaHash);
+    event ModuleEnabled(address safe, address module, bool allowRootAccess);
+    event ModuleDisabled(address safe, address module);
+
+    // Errors
+    error ModuleRequiresRootAccess(address sender);
+    error MoudleNotEnabled(address module);
+    error ModuleEnabledOnlyForRootAccess(address module);
+    error ModuleAccessMismatch(address module, bool requiresRootAccess, bool providedValue);
+    error ActionExecutionFailed(address safe, bytes32 metaHash, uint256 index);
+    error RootAccessActionExecutionFailed(address safe, bytes32 metaHash);
+
+    constructor(address initalOwner) {
+        _transferOwnership(initalOwner);
+    }
+
+    modifier onlyEnabledModule(ISafe safe) {
+        if (!enabledComponents[address(safe)][msg.sender].enabled) {
+            revert MoudleNotEnabled(msg.sender);
+        }
+        _;
+    }
+
+    /**
+     * @notice TODO
+     * @param safe TODO
+     * @param transaction TODO
+     * @return data TODO
+     */
+    function executeTransaction(
+        ISafe safe,
+        SafeTransaction calldata transaction
+    ) external override onlyEnabledModule(safe) returns (bytes[] memory data) {
+        // TODO: Check for re-entrancy attacks
+        // TODO: Validate metahash
+
+        if (ISafeProtocolModule(msg.sender).requiresRootAccess()) {
+            revert ModuleRequiresRootAccess(msg.sender);
+        }
+
+        if (enabledComponents[address(safe)][msg.sender].rootAddressGranted) {
+            revert ModuleEnabledOnlyForRootAccess(msg.sender);
+        }
+
+        data = new bytes[](transaction.actions.length);
+        for (uint256 i = 0; i < transaction.actions.length; ++i) {
```

Comment:
I would recommend to encode this into a multisend call (at least if there is more than 1 action), this is way more gas efficient.

Reply:
When using [multiSend](https://github.com/safe-global/safe-contracts/blob/6434676c47fe50fa0d21f56b26747aa1216504ef/contracts/libraries/MultiSend.sol#L30), how can we then return the data for each individual action?

Reply:
Lets keep it as is for now even so it costs more gas, as the other stuff is an implementation overhead. (Maybe we can open an issue to evaluate the complexity and gas impact in the future)

Reply:
Issue: #15